crates/tools/sim_chat/src/main.rs

```diff
@@ -0,0 +1,164 @@
+//! SimChat Command Line Utility / Manual Testing CLI
+
+extern crate linefeed;
+extern crate regex;
+extern crate url;
+use lib3h_sim_chat::ChatEvent;
+use regex::Regex;
+use url::Url;
+
+use lib3h::{
+    dht::mirror_dht::MirrorDht,
+    engine::{ghost_engine::GhostEngine, RealEngineConfig},
+    transport::{memory_mock::transport_memory::TransportMemory, TransportWrapper},
+    transport_wss::TlsConfig,
+};
+use lib3h_sodium::SodiumCryptoSystem;
+
+fn engine_builder() -> GhostEngine<'static> {
+    let network_transport = TransportWrapper::new(TransportMemory::new());
+    let crypto = Box::new(SodiumCryptoSystem::new());
+    let config = RealEngineConfig {
+        tls_config: TlsConfig::Unencrypted,
+        socket_type: "mem".into(),
+        bootstrap_nodes: vec![],
+        work_dir: String::new(),
+        log_level: 'd',
+        bind_url: Url::parse(format!("mem://{}", "test_engine").as_str()).unwrap(),
+        dht_gossip_interval: 100,
+        dht_timeout_threshold: 1000,
+        dht_custom_config: vec![],
+    };
+    let dht_factory = MirrorDht::new_with_config;
+
+    GhostEngine::new(
+        "test_engine",
+        crypto,
+        config,
+        dht_factory,
+        network_transport,
+    )
+    .unwrap()
+}
+
+fn main() {
+    let rl =
+        std::sync::Arc::new(linefeed::Interface::new("sim_chat").expect("failed to init linefeed"));
+
+    rl.set_report_signal(linefeed::terminal::Signal::Interrupt, true);
+    rl.set_prompt("SimChat> ")
+        .expect("failed to set linefeed prompt");
+
+    let rl_t = rl.clone();
+    let mut cli = lib3h_sim_chat::SimChat::new(
+        engine_builder,
+        Box::new(move |event| {
+            match event {
+                ChatEvent::JoinSuccess { channel_id, .. } => {
+                    rl_t.set_prompt(&format!("#{}> ", channel_id).to_string())
+                        .expect("failed to set linefeed prompt");
```

Comment:
Should we have a channel HashSet? It might be cool to let them join any number of channels, and if they join one they've previously already join, it just promotes that one to the prompt (i.e. the one outgoing messages will be sent on), and they will see incoming messages on all channels.

alternately, when they do join and are already part of a channel we could automatically send a LeaveSpace